### PR TITLE
PERF-2089 Update ^RandomDate Generator to accept other generators as input

### PR DESCRIPTION
**Pull request by jimoleary**
**dateStringParserGenerator** calls **stringGenerator**. As a result, it supports all the generators (even if they don't make sense like ^IP).

### CODE REVIEW

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [inline src/value_generators/src/DocumentGenerator.cpp:157]: I feel like max_date should be a lot higher. I know it's what we were already using, but it feels quite stark now.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [inline src/value_generators/src/DocumentGenerator.cpp:686]: Can I use {^Now: {}} as the min value? Does dateGenerator case to an int64?

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [inline src/value_generators/src/DocumentGenerator.cpp:699]: With the extra parameterization and generators, do you want to make kinds based on all the int distribution generators?

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [inline src/value_generators/src/DocumentGenerator.cpp:1148]: This looks nice.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [reply src/value_generators/src/DocumentGenerator.cpp:686]: It looks like it can. ^Now is in dateGenerator. Does the system automatically do the conversion?

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [inline src/value_generators/src/DocumentGenerator.cpp:1229]: Oh, we need to be careful here. So this will evaluate only once. Whatever string you get the first time is what you have the whole time. I'm going back and forth with myself over what makes the most sense. 

Could we make a UniqueGenerator<int64_t> that wraps a UniqueGenerator<string>?

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [inline src/value_generators/src/DocumentGenerator.cpp:1278]: Should we order these by expected frequency? Like dateStringParser is probably most likely, then dateTimeGenerator. I'm thinking about all the try catches. I wonder if there is a way around it. It's a valid way to do it.

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (dalyd) [inline src/value_generators/test/DocumentGeneratorTestCases.yml:598]: This is really cool!

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (dalyd) [inline src/value_generators/test/DocumentGeneratorTestCases.yml:600]: I missed it, do we check that max > min? With generators it will be easy to construct a case in which that is not true.

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (dalyd) [inline src/value_generators/test/DocumentGeneratorTestCases.yml:605]: Typo in name

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (dalyd) [inline src/value_generators/test/DocumentGeneratorTestCases.yml:628]: I love this!

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (dalyd) [inline src/value_generators/test/DocumentGeneratorTestCases.yml:642]: I can get a random generator that is (2021-01-01, ^Now). Can I somehow specify (^Now - 30 days, ^Now). I would like that, but can manage with the first. Obviously not that syntax.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (jimoleary) [reply src/value_generators/src/DocumentGenerator.cpp:157]: 2050?

Comment (dalyd) [reply src/value_generators/src/DocumentGenerator.cpp]: 2150?

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [reply src/value_generators/src/DocumentGenerator.cpp]: That's cool. Nice job.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [reply src/value_generators/src/DocumentGenerator.cpp]: Yeah, keep the scope reasonable. I could imagine though having a set range (today to end of year), but not wanting that to be uniformly distributed.

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (dalyd) [reply src/value_generators/test/DocumentGeneratorTestCases.yml:661]: Cool.  Sounds good.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (dalyd) [reply src/value_generators/src/DocumentGenerator.cpp:1241]: Could you make sure to include comments with the documentation for this? That may only be the test cases, but still.

**Review by dalyd — APPROVED** [review]
As I'm out next week, I'm going to approve now. I'll leave the rest with Ryan. 
One last request: Could we include some of these generators in a workload in the Docs directory? In Generators.yml?

File: src/value_generators/src/DocumentGenerator.cpp
Comment (jimoleary) [reply src/value_generators/src/DocumentGenerator.cpp:1241]: Is the fact that  a ```{^Now: {}}``` evaluated once a surprise? 

If so then it may make sense to create a lazy generator for these two types.

Comment (dalyd) [reply src/value_generators/src/DocumentGenerator.cpp:1241]: It is. I had proposed another generator of {^Once: {# Some other generator}} to allow explicitly doing that because there are use cases for it.

**Review by rtimmons — APPROVED** [review]
I'm OOO and can't look deeper until Tuesday. I took a quick look and I saw nothing would prevent merging, so if this needs to go in sooner I'm happy if David and the CI tests are happy.

**Comment by jimoleary** [comment]
@rtimmons I'm happy to wait for your review. And it will give me a bit of time to look at lazy evaluating generators.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (jimoleary) [reply src/value_generators/src/DocumentGenerator.cpp:1241]: ok, I'll have a look at lazy evaluating generators. To be correct / consistent all the non-constant cases should also be lazy evaluated (i.e.^RandomInt etc.... ).

There wil be a side effect though that the min / max check could pass or fail depending on what values are generated.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (jimoleary) [inline src/value_generators/src/DocumentGenerator.cpp:685]: Is this the correct approach here?

File: src/value_generators/src/DocumentGenerator.cpp
Comment (jimoleary) [inline src/value_generators/src/DocumentGenerator.cpp:739]: The error detection happens in evaluate so the test runner needed to be changed to take this into account.

File: src/value_generators/test/DocGenYamlTestCaseRunner.cpp
Comment (jimoleary) [inline src/value_generators/test/DocGenYamlTestCaseRunner.cpp:72]: evaluate for ^RandomDate can throw the exception.

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (jimoleary) [inline src/value_generators/test/DocumentGeneratorTestCases.yml:543]: max date changed in the last commit which has a knock on effect on the dates generated.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:157]: plz use `=` or curlies to make it clear these are static members not forward references.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:156]: c++-ism - the `static` isn't needed here

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:169]: 💅🏽 plz use `camelCase` - `const ptime& defaultTime`

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [reply src/value_generators/src/DocumentGenerator.cpp:685]: I'm not sure what the intent here is and why the `std::swap` does what you want? I don't usually see `swap` used with refs - usually it's with value types or values. This could be totally kosher but I for one could use a comment here about why this is the right thing.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:690]: 💅🏽 plz remove the commented-out logs

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:709]: meh - `private` unless you intend these to be subclasses?

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:704]: so `_gen->eval()` returns `double` which we cast to `long long` but the expected return type is `int64`? I think this is working by casting twice which is why you needed the separate `auto millis` declaration.

I think you can use a single cast:

```c++
int64_t evaluate() override {
  return (int64_t) _gen->eval();
}
```

Also consider naming this `CastDoubleToInt64Generator` or similar.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:1201]: we could make a separate `UnknownParserException` if that would help? Relying on exception messages for logic has bitten me a number of times in the past (trying to make exceptions easier to read has caused bugs). If that doesn't seem worthwhile or like it will work plz add a comment wherever it adds the "Unknown parser" string letting future maintainers know about `isUnknownParser`. (Obv same to the format one below)

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:1234]: why not `return millis` here?

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:1222]: another `camelCase` reminder

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:1338]: Does a lot of this of this go away if we don't allow mixed types in the min/max fields? I.e. if you want a literal min/max rather than generated simply require the use of `^Verbatim`? What you have here is clever and well-written, but I wonder if it's complexity worth having.

File: src/value_generators/test/DocumentGeneratorTestCases.yml
Comment (rtimmons) [reply src/value_generators/test/DocumentGeneratorTestCases.yml:612]: > it's still easy to construct a case where it aribrarily fails depending on the inputs.

This seems a bit...fraught. Would it make sense for there to be a "min" and "max" member-function on the date generators that returns the min/max values that the generator could ever produce and then apply the min/max checking logic at construction time? This is obviously rather tricky with `^Choose` and perhaps others but maybe that's a good thing - meaning maybe it shouldn't be possible to internally use something other than `DateGen` (isA `Generator<boost::time_point>` or whatever the boost type is)?

**Review by rtimmons — APPROVED** [review]
Looks good - I like seeing the recursion machinery working out reasonably well for this.

My only hesitation is around the `std::swap` which may be correct but just not paging into my C++ memory banks on correct idioms.

Oh ya one more thing--
Adding a couple `ThenReturn` values for repeated executions may be helpful to ensure the runtime evaluation of certain things. E.g. I can imagine

```yaml
GivenTemplate:
  date: {^RandomDate: {min: {^Choose: {from: ["1500-01-01", "2020-02-16"]}, max: "2021-02-16"}
ThenReturns: [{date: foo}, {date: bar}] # ideally where bar couldn't have been from a choice of 1500 being the min
```

I'm not feeling clever enough to come up with a better example, but right now I think it's possible to introduce a regression that reverts back to instantiation-time evaluation and these test may not protect against that. If I'm wrong I'm happy to leave the tests as they are and not address this explicitly.

**Comment by rtimmons** [comment]
Also this whole generators-of-generators thing is starting to feel like we're dangerously close to implementing a lisp. I'm not opposed to that per see but we could decide to lean into it more if we want to be more and more clever over time. Just some :philosoraptor: commentary.

**Review by jimoleary — COMMENTED** [review]
> Also this whole generators-of-generators thing is starting to feel like we're dangerously close to implementing a lisp. I'm not opposed to that per see but we could decide to lean into it more if we want to be more and more clever over time. Just some :philosoraptor: commentary.

The 'killer' feature that this work allows is **_{^RandomDate: {min: {^Now:{}}}}_** where Now is the current time (rather than the first invocation time).

It seemed natural to me to make all the other generators evaluate at runtime also. It didn't add a huge amount of extra work.

In terms of min / max validation, your suggestion of adding a set of min / max methods to return the correct values to allow for validation has merit although unfortunately ^Now would require extra thought. Should I raise a new ticket to investigate?

There is an issue with boost date parsing that needs a bit of investigation, I've raised PERF-2153 to track this.

File: src/value_generators/src/DocumentGenerator.cpp
Comment (rtimmons) [inline src/value_generators/src/DocumentGenerator.cpp:710]: I think this can introduce use-after-move problems:

```c++
UniqueGenerator<std::string> og = ....;
DateStringParserGenerator other {og};
og->evaluate(); // 💥
```

I realize mostly academic for this code, but see if you can just pass in an rvalue and do the `std::move` at the callsite:

```c++
class ... {
     explicit DateStringParserGenerator(UniqueGenerator<std::string>&& generator)
     : _gen{generator} {}
}
...

UniqueGenerator<std::string> og = ....;
DateStringParserGenerator other {std::move(of)};
og->evaluate(); // still runtime 💥 but compiler can detect it
```

Ugh my goodness C++ makes it far too easy to footgun.

Comment (jimoleary) [reply src/value_generators/src/DocumentGenerator.cpp:710]: Thanks, I'll change it. I did think of doing the move at the call site, but thought it would be easier to do it in the receiver.

Comment (jimoleary) [reply src/value_generators/src/DocumentGenerator.cpp:710]: It looks like there is some oddness in my linux dev environment, the code above fails to compile on ubuntu (it works on MacOS (thanks @rtimmons). I will investigate my dev env a bit more and come back.

**Comment by jimoleary** [comment]
evergreen merge

**Pull request merged**
